### PR TITLE
#4839 Configurable local plugins folder feature.

### LINKED ISSUE #4839: Configure local plugins folder
# This is a Feature Proposal

## Description

Right now, when creating small local plugins, the only current option is to create a folder named `.serverless_plugins` to store it.

It would be nice to be able to configure the local plugins folder. This is good for projects with a well defined folder structure, that must then have an enforced folder that breaks the mould. That is specially true for _dot_ prefixed folders, which are usually used for working environment metadata, and commonly _gitignored_.

Behaviour would be exactly the same as today, it would only change the folder from `.serverless_plugins` to the configured one.

```yaml
service:
  name: my-serverless-stuff

# we could have a simple new property...
localPluginsFolder: './sls/plugins'
plugins:
  - plugin-one
  - plugin-two
...
```

```yaml
service: my-serverless-stuff  # fallback!

# or we could have an enhanced plugin object, with fallback into the "import", 
# just as "service" does with "name" 
plugins:
  localFolder: './sls/plugins'
  import:
    - plugin-one
    - plugin-two
...
```
## Additional Data

* https://serverless.com/framework/docs/providers/aws/guide/plugins/

### PR DESCRIPTION
**Pull request by GuruRAM**
## What did you implement:

Closes #4839 Configurable local plugins folder

## How did you implement it:
PluginManager.js was extended with the additional method `parsePluginsObject `which parse the plugins definition and return `localPath` and `modules` (plugins). If path can't be retrieved the default path is returned (`.serverless_plugins`).
PackageService.js was extended and now `getExcludes `method now adds localPath to excludes.

## How can we verify it:
Unit tests were added for both classes.
Can be tested with the following yml file:
`service: localplugin
plugins:
  localPath: './sls_plugins'
  modules:
    - my-plugin-test
provider:
  name: aws
  runtime: nodejs6.10
functions:
  hello:
    handler: handler.hello
`

## Todos:

- [x] Write tests
- [ ] Write documentation - when the logic is settled, documentation will be updated
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO - Because the rollback logic is present

### CODE REVIEW

File: lib/plugins/package/lib/packageService.js [hunk 1]
```diff
@@ -26,9 +26,13 @@ module.exports = {
 
   getExcludes(exclude) {
     const packageExcludes = this.serverless.service.package.exclude || [];
-
+    // add local service plugins Path
+    const pluginsLocalPath = this.serverless.service.plugins &&
+      this.serverless.service.plugins.localPath;
+    const localPathExcludes = pluginsLocalPath &&
+      typeof pluginsLocalPath === 'string' ? [pluginsLocalPath] : [];
```
Comment (HyperBrain) [inline lib/plugins/package/lib/packageService.js]: You should use lodash's `_.isString()` here

File: lib/classes/PluginManager.js [hunk 2]
```diff
@@ -130,16 +130,32 @@ class PluginManager {
...
+
+  parsePluginsObject(servicePlugs) {
+    let localPath = (this.serverless && this.serverless.config &&
+      this.serverless.config.servicePath) &&
+      path.join(this.serverless.config.servicePath, '.serverless_plugins');
+    let modules = [];
+
+    if (Array.isArray(servicePlugs)) {
+      modules = servicePlugs;
+    } else if (servicePlugs && Array.isArray(servicePlugs.modules)) {
```
Comment (HyperBrain) [inline lib/classes/PluginManager.js]: Use lodash's `_.isArray()` here

File: lib/classes/PluginManager.js [hunk 3]
```diff
@@ -130,16 +130,32 @@ class PluginManager {
...
+      modules = servicePlugs.modules;
+      localPath = servicePlugs.localPath &&
+        typeof servicePlugs.localPath === 'string' ? servicePlugs.localPath : localPath;
```
Comment (HyperBrain) [inline lib/classes/PluginManager.js]: `_.isString()`

File: lib/classes/PluginManager.test.js [hunk 4]
```diff
@@ -779,6 +779,64 @@ describe('PluginManager', () => {
     });
   });
 
+  describe('#parsePluginsObject()', () => {
+    const parsePluginsObject = (servicePlugins, expectedResult) => {
```
Comment (HyperBrain) [inline lib/classes/PluginManager.test.js]: You should name this helper function `parsePluginsObjectAndVerifyResult`. Then it is clear that you do not call the original `parsePluginsObject` function directly in the tests below, but combine verification and call in one function.

File: lib/plugins/package/lib/packageService.js (hunk 1 above)
Comment (GuruRAM) [reply lib/plugins/package/lib/packageService.js]: Done!

File: lib/classes/PluginManager.test.js (hunk 4 above)
Comment (GuruRAM) [reply lib/classes/PluginManager.test.js]: Thank you, I have renamed the name to make it more clear.

File: lib/classes/PluginManager.js (hunk 3 above)
Comment (GuruRAM) [reply lib/classes/PluginManager.js]: I have replaced native calls to lodash isObject, isString and isArray.

**Comment by GuruRAM** [comment]
@HyperBrain I have fixed the changes regarding lodash and function name

**Comment by horike37** [comment]
@GuruRAM @HyperBrain 
I tested locally with the plugin command but unfortunately the command is broken by this functionality. As the specification, `sls plugin install` automatically add installed plugins array under `plugins` section in serverless.yml so we can't specify object value there.

I think would be good to move `localPath` under provider section something like this.

```
provider:
  pluginLocalPath: './sls_plugins'

plugins:
  - your-own-local-plugin
  - community-plugin-1
  - community-plugin-2
```

What do you think?

**Comment by HyperBrain** [comment]
@horike37 What about moving the plugins declaration to a dual object like we did with the service declaration? So that legacy components like the plugins command can use it too - or adjust the plugins command to use the new semantics.

Moving the `localPath` property outside of the plugins property is imo not a good solution, because it only defines the local path for the plugins, not the provider or anything else.

**Comment by GuruRAM** [comment]
@horike37 @HyperBrain Imo the plugins declaration should be in one place and not spread between different sections.

The only thing I'm not comfortable with is the `parsePluginsObject` logic will partially reside in `packageService.js`, `PluginManager.js` and `plugin/install.js` but it can be moved here for example: `plugins\plugin\utils.js` (not sure about the place though).

**Comment by horike37** [comment]
@HyperBrain @GuruRAM 
That makes sense. I think it would be a better approach than mine :+1: 

Currently, adding and removing plugins array are done in the following place.
You can realize the solution by adding if-check whether is an array or object.
https://github.com/serverless/serverless/blob/master/lib/plugins/plugin/install/install.js#L132
https://github.com/serverless/serverless/blob/master/lib/plugins/plugin/uninstall/uninstall.js#L100

**Comment by GuruRAM** [comment]
@horike37 @HyperBrain I had to change the implementation of `yamlAstParser`.
The main problem was with branch overriding during add / remove operations.
If the remove plugin2 operation will be performed on the following file:
```
plugins:
  localPath: 'test'
  modules:
    -  plugin1
    -  plugin2
```
It will be transformed to:
```
plugins:
  modules:
    -  plugin1
```
Because the current implementation replaces the the root element (plugins in our case) with the new one and copy only an array into the new root element:
```
    let newObject = JSON.stringify(newArray);
    _.forEach(pathInYmlArrray.reverse(), (v) => {
      newObject = `{"${v}":${newObject}}`;
    });
```
I tried different implementations:

1. I tried to replace only the array element without replacing the whole root element. But it was rather complicated and required some YAML hacks (like manual indent management).
2. **Implemented solution**.  The root element is replaced but the new element contains all the nodes of the old one. I have created `constructPlainObject`  method that constructs configuration plain object from the AST Node.
Unit tests are extended accordingly.

**Comment by horike37** [comment]
@GuruRAM 
You don't have to change codes inside `yamlAstParser` since you can change second node with the current functionality of `yamlAstParser.addNewArrayItem` and `removeExistingArrayItem` something like this

```
return yamlAstParser.removeExistingArrayItem(serverlessFilePath, 'plugins.modules', this.options.pluginName);
```

```
return yamlAstParser.addNewArrayItem(serverlessFilePath, 'plugins.modules', this.options.pluginName);
```
Therefore, You need to check whether to be array or object under `plugins` element and you will apply `plugins.modules` to the second argument of these methods if it is object.

**Comment by GuruRAM** [comment]
@horike37 Initially I tried it. The current functionality will remove localPath property from the plugins object. It can be easily tested by extending existing `yamlASTParser.test.js`

Here is the sample test:
```
    it.only('insertion into complex object test', () => {
      const yamlFilePath = path.join(tmpDirPath, 'test.yaml');

      writeFileSync(yamlFilePath, {
        plugins: {
          modules: ['foo'],
          localPath: 'test'
        }
      });
      return expect(yamlAstParser.addNewArrayItem(yamlFilePath, 'plugins.modules', 'bar'))
        .to.be.fulfilled.then(() => {
          const yaml = readFileSync(yamlFilePath, 'utf8');
          expect(yaml).to.be.deep.equal({
            plugins: {
              modules: ['foo', 'bar'],
              localPath: 'test'
            }});
          });
    });
```

The output will be:
```
{
            plugins: {
              modules: ['foo', 'bar']
            }
}
```

And `removeExistingArrayItem` behaves the same way it will remove localPath property since its recreating the whole path from the leaf till the root but only the current leaf will be included into the branch. There is the example:
```
    it.only('deletion from complex object test', () => {
      const yamlFilePath = path.join(tmpDirPath, 'test.yaml');

      writeFileSync(yamlFilePath, {
        plugins: {
          modules: ['foo', 'bar'],
          localPath: 'test'
        }
      });
      return expect(yamlAstParser.removeExistingArrayItem(yamlFilePath, 'plugins.modules', 'bar'))
        .to.be.fulfilled.then(() => {
          const yaml = readFileSync(yamlFilePath, 'utf8');
          expect(yaml).to.be.deep.equal({
            plugins: {
              modules: ['foo'],
              localPath: 'test'
            }});
          });
    });
```

**Comment by horike37** [comment]
@GuruRAM 
Ah, sorry. I missed the case which those is mixed together, so we need to extend the `yamlAstParser` so that the case can be worked as you stated above comment.

**Comment by GuruRAM** [comment]
@horike37 no problems at all! I have extended the functionality and unit tests already so these cases are also covered. Can you please check unit tests and the implementation.

File: lib/classes/PluginManager.js [hunk 5]
```diff
@@ -54,9 +54,9 @@ class PluginManager {
     let pluginProvider = null;
     // check if plugin is provider agnostic
     if (pluginInstance.provider) {
-      if (typeof pluginInstance.provider === 'string') {
+      if (_.isString(pluginInstance.provider)) {
         pluginProvider = pluginInstance.provider;
-      } else if (typeof pluginInstance.provider === 'object') {
+      } else if (_.isObject(typeof pluginInstance.provider)) {
```
Comment (HyperBrain) [inline lib/classes/PluginManager.js]: The typeof is superfluous and should be removed: `_.isObject(pluginInstance.provider)`

Comment (GuruRAM) [reply lib/classes/PluginManager.js]: @HyperBrain fixed, refactoring type.

**Comment by HyperBrain** [comment]
@horike37 Looks good for me now. Can you do a final review regarding the plugin command and the unit tests?

**Comment by GuruRAM** [comment]
@horike37 @HyperBrain Cool, thank you!
I have updated the documentation for plugins. I have also updated #installing-plugins to reflect the second available format for the plugins section that is supported.

**Comment by horike37** [comment]
@GuruRAM 
Thank you for the updates! Looks like perfect :smile: 
Currently, we are in merge freeze term for preparation of v1.27.0 as the next version.
This will be shipped in v1.28.0 release!

**Comment by HyperBrain** [comment]
@horike37 As you did the important plugin command verifications, could you merge this one in as soon as you are ok with it?

**Comment by GuruRAM** [comment]
@horike37 I have merged the latest master to the pull request branch.

**Comment by horike37** [comment]
@GuruRAM 
Thanks for pinging! just merged

**Pull request merged**